src/thor/trace_route_action.cc

```diff
@@ -490,29 +504,76 @@ thor_worker_t::map_match(Api& request) {
             // when handling multi routes, orsm serializer need to know both the
             // matching_index(route_index) and the waypoint_index.
             Location* origin_location =
-                options.mutable_shape(leg_origin_iter - match_results.cbegin());
+                options.mutable_shape(origin_match_result - match_results.cbegin());
             origin_location->set_route_index(route_index);
             origin_location->set_shape_index(way_point_index);
 
             Location* destination_location =
-                options.mutable_shape(leg_destination_iter - match_results.cbegin());
+                options.mutable_shape(destination_match_result - match_results.cbegin());
             destination_location->set_route_index(route_index);
             destination_location->set_shape_index(++way_point_index);
 
-            add_path_edge(&*origin_location, *leg_origin_iter);
-            add_path_edge(&*destination_location, *leg_destination_iter);
+            // we fake up something that looks like the output of loki
+            add_path_edge(&*origin_location, *origin_match_result);
+            add_path_edge(&*destination_location, *destination_match_result);
+
+            // mark the beginning and end of the edges on the path for this leg (inclusive)
+            auto leg_begin = std::next(std::get<0>(edge_group), last_edge_index);
+            auto leg_end = path_edge_itr;
+
+            // "handling partial edges will be the death of us" -- Confucius 504 BC
```

Comment:
from here down is pretty much the important part of fixing the map matching elapsed duration tracking. this comment pretty well describes whats going on as well as the other comments below it